gufe/tests/test_protocolunit.py

```diff
@@ -62,84 +82,141 @@ def test_key_differs(self):
 
-            shared = Path("shared") / str(unit.key)
-            shared.mkdir(parents=True)
+        shared = Path("shared") / str(unit.key)
+        shared.mkdir(parents=True)
 
-            scratch = Path("scratch") / str(unit.key)
-            scratch.mkdir(parents=True)
+        scratch = Path("scratch") / str(unit.key)
+        scratch.mkdir(parents=True)
 
-            ctx = Context(shared=shared, scratch=scratch, stderr=None, stdout=None)
+        ctx = Context(
+            shared_storage=shared_storage,
+            permanent_storage=permanent_storage,
+            dag_label="test",
+            unit_label=unit.key,
+            scratch=scratch_storage,
+        )
 
-            with pytest.raises(ExecutionInterrupt):
-                unit.execute(context=ctx, an_input=3)
+        with pytest.raises(ExecutionInterrupt):
+            unit.execute(context=ctx, an_input=3)
 
-            u: ProtocolUnitResult = unit.execute(context=ctx, an_input=2)
+        u: ProtocolUnitResult = unit.execute(context=ctx, an_input=2)
 
-            assert u.outputs == {"foo": "bar"}
+        assert u.outputs == {"foo": "bar"}
 
-    def test_execute_KeyboardInterrupt(self, tmpdir):
-        with tmpdir.as_cwd():
-            unit = DummyKeyboardInterruptUnit()
+    def test_execute_KeyboardInterrupt(self, scratch_storage, permanent_storage, shared_storage):
+        unit = DummyKeyboardInterruptUnit()
 
-            shared = Path("shared") / str(unit.key)
-            shared.mkdir(parents=True)
+        shared = Path("shared") / str(unit.key)
+        shared.mkdir(parents=True)
 
-            scratch = Path("scratch") / str(unit.key)
-            scratch.mkdir(parents=True)
+        scratch = Path("scratch") / str(unit.key)
+        scratch.mkdir(parents=True)
 
-            ctx = Context(shared=shared, scratch=scratch, stderr=None, stdout=None)
+        ctx = Context(
+            shared_storage=shared_storage,
+            permanent_storage=permanent_storage,
+            dag_label="test",
+            unit_label=unit.key,
+            scratch=scratch_storage,
+        )
 
-            with pytest.raises(KeyboardInterrupt):
-                unit.execute(context=ctx, an_input=3)
+        with pytest.raises(KeyboardInterrupt):
+            unit.execute(context=ctx, an_input=3)
 
-            u: ProtocolUnitResult = unit.execute(context=ctx, an_input=2)
+        u: ProtocolUnitResult = unit.execute(context=ctx, an_input=2)
 
-            assert u.outputs == {"foo": "bar"}
+        assert u.outputs == {"foo": "bar"}
 
     def test_normalize(self, instance):
         thingy = instance.key
 
         assert thingy.startswith("DummyUnit-")
         assert all(t in string.hexdigits for t in thingy.partition("-")[-1])
+
+
+class TestContext:
+    """Test the Context class context manager functionality."""
+
+    def test_context_manager_enter_exit(
+        self, scratch_storage, shared_storage: MemoryStorage, permanent_storage: MemoryStorage
```

Comment:
we should instead (or in addition) test this with file storage. right now the cleanup isn't covered in tests: https://app.codecov.io/gh/OpenFreeEnergy/gufe/pull/690?src=pr&el=tree&filepath=gufe%2Fprotocols%2Fprotocolunit.py&utm_medium=referral&utm_source=github&utm_content=checks&utm_campaign=pr+comments&utm_term=OpenFreeEnergy#diff-Z3VmZS9wcm90b2NvbHMvcHJvdG9jb2x1bml0LnB5